### ISSUE TITLE
[🚀 Feature]: [dotnet] Road to be AOT friendly

### ISSUE THREAD

**Issue by nvborisenko** [SeleniumHQ/selenium#14480]
### Feature and motivation

Any consumer of Selenium library in .NET cannot use it within publishing his application as [Native AOT Deployment](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net7%2Cwindows).

The detailed affected parts are coming...

**Selenium Manager** ✅ **v4.26**
- [x] Selenium Manager - #14481 

**W3C WebDriver** ✅ **v4.26**
- [x] New Session
- [x] ... - #14574

**W3C BiDi** ✅ **v4.28**
- [x] ... https://github.com/SeleniumHQ/selenium/pull/15162

**CDP (DevTools)**
- [ ] Generate JsonSerializerContext for all DTO models
- [ ] ...

**Fine tuning**
- Address all analyzer warnings
- [ ] ...

### Usage example

Create console app, publish as native AOT, it fails at runtime.

**Comment by nvborisenko** [comment]
Blocked by  https://github.com/bazelbuild/rules_dotnet/issues/443

**Comment by RenderMichael** [comment]
I’m interested in this feature for running Selenium. Often times, the roads to AOT compatibility leads to improvements across the board, even for non-AOT scenarios.

**Comment by NCLnclNCL** [comment]
> Blocked by [bazelbuild/rules_dotnet#443](https://github.com/bazelbuild/rules_dotnet/issues/443)

when it will be fixed

**Comment by nvborisenko** [comment]
> Blocked by https://github.com/bazelbuild/rules_dotnet/issues/443

Fixed, waiting release soon.

> Often times, the roads to AOT compatibility leads to improvements across the board, even for non-AOT scenarios.

Who knows potential issues... , I cannot prepare decomposed tasks to archive the goal. Our mission here is to adopt existing codebase to be ready for AOT.

The technical enabler is resolved, so the list of impacted areas will be be growing so fast soon.

**Comment by nvborisenko** [comment]
@RenderMichael are you able to test the preview version of Selenium which seems is compatible with AOT (excepting CDP)? Your input would be valuable. I can share nuget package file here (built locally).

**Comment by RenderMichael** [comment]
@nvborisenko I would be happy to test out the AOT compatibility of Selenium :)

**Comment by nvborisenko** [comment]
[Selenium.WebDriver.4.26.0-nightly202409202352.zip](https://github.com/user-attachments/files/17422890/Selenium.WebDriver.4.26.0-nightly202409202352.zip)

@RenderMichael  This is zip, which contains nupkg inside. I built it locally, and hoping you know what to do with it.

**Comment by nvborisenko** [comment]
@RenderMichael @NCLnclNCL please don't forget to share results. Thanks guys.

**Comment by RenderMichael** [comment]
~~@nvborisenko Unfortunately the app crashed as soon as I tried running the native .exe file:~~

![image](https://github.com/user-attachments/assets/6d9f3855-2383-4e66-aa08-928969302888)

~~I can't give more detailed feedback without building the library locally. Are there any instructions I can follow to build the .NET bindings locally? A while back I tried to do so following [this guide](https://jimevansmusic.blogspot.com/2020/04/setting-up-windows-development.html) but ran into some blockers (I no longer remember what they were).~~

Edit: it was my error

**Comment by RenderMichael** [comment]
This was my program for the above. Completely ideomatic, no obscure Selenium features or anything.

```csharp
using OpenQA.Selenium.Firefox;

var options = new FirefoxOptions
{
    AcceptInsecureCertificates = true,
};

using (var driver = new FirefoxDriver(options))
{
    await driver.Navigate().GoToUrlAsync("https://www.google.com");

    string url = driver.Url;
    if (!url.Contains("google"))
    {
        throw new Exception("URL does not contain google: " + url);
    }
}

Console.ReadKey();
```

**Comment by nvborisenko** [comment]
@RenderMichael did you use nupkg file I provided above? Or you just referenced to dll from that package?

**Comment by NCLnclNCL** [comment]
> @RenderMichael @NCLnclNCL please don't forget to share results. Thanks guys.

it works very well, the file size is also very light

**Comment by nvborisenko** [comment]
@NCLnclNCL did you try it on your "production" project? Or just on simple "Hello World"?

**Comment by RenderMichael** [comment]
@nvborisenko Ah that was from a fluke antivirus mis-configuration on my part, I didn't release there was supposed to be a `selenium-manager` folder alongside the program, and the `TypeInitalizationException` in my stack trace threw me off. Sorry!

The `Hello World` is working great for me now, I will try with a more serious program. Thanks!

**Comment by RenderMichael** [comment]
@nvborisenko I found a few non-JSON related improvements we can make to the AOT story, I opened https://github.com/SeleniumHQ/selenium/pull/14637

**Comment by RenderMichael** [comment]
I opened an issue to improve the exception experience of what initially confused me: https://github.com/SeleniumHQ/selenium/pull/14639

**Comment by nvborisenko** [comment]
Seems success, we don't have reported issues related to json serialization. Thanks all, who participated in this. I deprioritized CDP, we are more interested in BiDi.

**Comment by RenderMichael** [comment]
I think success for this can be measured by adding `<IsAotCompatible>` to `true`, and add the `[RequiresUnreferencedCode]` attribute for things that are not AOT-compatible. I can spin up a PR for the minimal changes required for this

**Comment by nvborisenko** [comment]
Please **no**, then we will create new issue like "Road to be trimming friendly".

**Comment by RenderMichael** [comment]
AOT is a subset of trimmable (or put another way, trim-friendly is a pre-requisite of AOT-friendly)

Besides, aside from JSON, I saw there were only 2 or 3 spots that need any changes

**Comment by nvborisenko** [comment]
I received a lot of issues privately related to json serialization. In major cases users used custom class and they provided an instance of it as `object`. Object is expected according API, and they got `JsonTypeInfo metadata for type '*' was not provided by TypeInfoResolver of type 'OpenQA.Selenium.CommandJsonSerializerContext'` exception. @RenderMichael provided elegant [fix](https://github.com/SeleniumHQ/selenium/commit/5c9d6833f822bfda1d9cde06bcf5231916a39649) (will be in 4.27). Seems we will forget about this issue for "classic" WebDriver commands. Thank you, Michael, great job!

**Comment by RenderMichael** [comment]
Happy to contribute!

I have a PR addressing every AOT concern aside from JSON. It was a lot simpler than I thought. #14637

**Comment by NCLnclNCL** [comment]
> Happy to contribute!
> 
> I have a PR addressing every AOT concern aside from JSON. It was a lot simpler than I thought. #14637

Hi,do you know anything about python, i want to convert python nodriver library to c#

**Comment by larsbloch** [comment]
Hi

My current goal is to migrate  my selenium project into a Native Aot application.

I get errors when i start a new DevToolsSessions

OpenQA.Selenium.WebDriverException: 'Unexpected error creating WebSocket DevTools session.'
InvalidOperationException: Reflection-based serialization has been disabled for this application. Either use the source generator APIs or explicitly configure the 'JsonSerializerOptions.TypeInfoResolver' property.

```
        var driver = new ChromeDriver();
        var devToolsSession = driver.GetDevToolsSession(); <- Error here
```

I am not familie with the internals of Selenium but maybe you can tell me if my project is a lost cause or if the problems are going to be handled by some of the hard work you seem to be doing in regards to AOT ?

**Comment by nvborisenko** [comment]
`GetDevToolsSession` is not compatible with AOT yet, please track `CDP` section in the original post.

**Comment by nvborisenko** [comment]
@RenderMichael do you want to invest into CDP (DevTools) AOT compatibility? I don't want, BiDi is a priority for me.

**Comment by RenderMichael** [comment]
@nvborisenko Per the tracking issue for depreciating CDP https://github.com/SeleniumHQ/selenium/issues/13994:

> We'll still allow access to lower level CDP functionality (see https://github.com/SeleniumHQ/selenium/issues/13990), but the high level approach will be removed.

I interpret this to mean we will still keep the `SendCommand` method which takes a `string commandName`, and `JsonNode commandParameters`, with the `JsonElement?` response. Basically remove the strongly typed commands.

For that reason, we shouldn't unnecessarily block AOT, especially since supporting it is so simple: https://github.com/SeleniumHQ/selenium/pull/15159/

**Comment by flucas1** [comment]
For my projects, I have found the flag JsonSerializerIsReflectionEnabledByDefault quite useful

````
		<JsonSerializerIsReflectionEnabledByDefault>false</JsonSerializerIsReflectionEnabledByDefault>
````

Even if there are parts of the code using reflection, at lest I know the Json part is reflection free... and therefore AOT and trimming safe
